### PR TITLE
refactor: turn on strict mode in Typescript configuration

### LINKED ISSUE #163: Turn on strictNullChecks in tsconfig.json
strictNullChecks is currently off right now, meaning there may be exceptions thrown by trying to access subkeys of `undefined` or `null` variables.

We should turn this on ASAP and fix all the compiler errors that arises.

### PR DESCRIPTION
**Pull request by karrui**
## Problem

Turn on strict mode that allows the Typescript compiler to alert us on undefined or nullable errors.

Various changes are made in the codebase to accommodate this change.

Also add better JSDocs on the functions I've come across

Closes #163

**Features**
- add vendor typings for `convict-format-with-validator` in `types/vendors`
- add vendor typing for `express-device` package in `types/vendors`
- add vendor typing for `bson-ext` package in `types/vendors`

**Improvements**
- Correct all errors that arise from the change to strict mode

**Dev Dependencies**
- bump `ts-jest` from 26.1.4 to 26.3.0: This removes the Typescript unsupported warning
- bump `@shelf/jest-mongodb` from 1.2.2 to 1.2.3: I thought this would allow us to not have to manually setup and teardown on `dbHandler.connect`, but that does not seem to be the case. No harm updating though so it stays.

### CODE REVIEW

**Comment by karrui** [comment]
The number of lines is a lie again, most of it is due to adding vendor typings to packages without types!!

**Comment by karrui** [comment]
I have absolutely no clue why testcafe fails to start....

File: src/app/models/field/attachmentField.ts [hunk 1]
```diff
@@ -30,7 +30,7 @@ const createAttachmentFieldSchema = () => {
   ) {
     const { webhook, responseMode } = this.parent()
 
-    if (responseMode === ResponseMode.Encrypt && webhook.url) {
+    if (responseMode === ResponseMode.Encrypt && webhook?.url) {
```
Comment (liangyuanruo) [inline src/app/models/field/attachmentField.ts:33]: were we at risk of a null pointer exception here?

File: src/app/models/form.server.model.ts [hunk 2]
```diff
@@ -416,6 +416,9 @@ const compileFormModel = (db: Mongoose): IFormModel => {
         path: 'admin',
         select: 'email',
       })
+      if (!data) {
```
Comment (liangyuanruo) [inline src/app/models/form.server.model.ts:419]: fairly nitpicky, one could prefer return with a ternary op in such cases as a matter of style. holding less state is always better.

```typescript
return data ? { ... } : null
```

File: src/app/modules/verification/verification.service.ts [hunk 3]
```diff
@@ -117,25 +122,26 @@ export const getNewOtp = async (
...
       `Wait for ${waitForSeconds} seconds before requesting for a new otp`,
       VfnErrors.WaitForOtp,
     )
   } else {
     const hashCreatedAt = new Date()
     const otp = generateOtp()
     const hashedOtp = await bcrypt.hash(otp, SALT_ROUNDS)
-    const signedData = formsgSdk.verification.generateSignature({
+
+    const signedData = formsgSdk.verification.generateSignature!({
```
Comment (liangyuanruo) [inline src/app/modules/verification/verification.service.ts:144]: is there a bug with the SDK type declarations?

Comment (karrui) [reply src/app/modules/verification/verification.service.ts:144]: Yes, it's fixed in a PR but it's not published yet I think, it's been a while, need to check for breaking changes too haha

File: src/app/modules/webhook/webhook.service.ts [hunk 4]
```diff
@@ -273,7 +273,7 @@ const getFormattedResponse = (
  */
 export const pushData = async (
   webhookUrl: WebhookParams['webhookUrl'],
-  submissionWebhookView: WebhookView,
+  submissionWebhookView: WebhookView | null,
```
Comment (liangyuanruo) [inline src/app/modules/webhook/webhook.service.ts:276]: ideally the null check should be done by the calling function, but we can always refactor later. it's much easier to spot non-canonical function signatures when the types are explicitly declared.

File: src/shared/util/file-validation.ts [hunk 5]
```diff
@@ -109,9 +109,11 @@ export const getInvalidFileExtensionsInZip = (
 
   // We wrap this checker into a closure because the data format
   // needs to be different for frontend vs backend.
-  const checkZipForInvalidFiles = async function (file: File | Buffer) {
+  const checkZipForInvalidFiles = async (
+    file: Blob | Buffer,
+  ): Promise<string[]> => {
     const zip = await JSZip.loadAsync(file)
-    const invalidFileExtensions = []
+    const invalidFileExtensions: (string | string[] | Promise<string[]>)[] = []
```
Comment (liangyuanruo) [inline src/shared/util/file-validation.ts:116]: lol. this probably needs a good refactor.

**Review by liangyuanruo — APPROVED** [review]
Thanks for getting this across the finish line; in retrospect strict mode probably should have been enabled at the start. I see plenty of non-null assertions being made, but I'll take what I can get.

File: src/app/modules/verification/verification.service.ts (hunk 3 above)
Comment (liangyuanruo) [reply src/app/modules/verification/verification.service.ts:144]: ok, that sounds like tech debt that we should address when you're back

**Comment by karrui** [comment]
> Thanks for getting this across the finish line; in retrospect strict mode probably should have been enabled at the start. I see plenty of non-null assertions being made, but I'll take what I can get.

I added lint rules for non null too, they are now warnings so it should get devs to be more aware too

File: src/app/models/field/attachmentField.ts (hunk 1 above)
Comment (karrui) [reply src/app/models/field/attachmentField.ts:33]: No, if it's encrypt mode it will have a webhooks object.

**Comment by mantariksh** [comment]
I'm working on fixing these merge conflicts, will merge once I'm done

**Pull request merged**